### PR TITLE
testing/pypy3: new aport

### PR DESCRIPTION
**Pull request by HorridBro**
https://github.com/HorridBro/AlpinePyPy3
PyPy3 build file

### CODE REVIEW

File: testing/pypy3/alpine.patch [hunk 1]
```diff
@@ -0,0 +1,187 @@
+diff -r 7f9407387dbf lib_pypy/_tkinter/tklib_build.py
```
Comment (jirutka) [inline testing/pypy3/alpine.patch]: Where is this patch from? Have you written it or copied from somewhere? What exactly does it fixes? It seems that it fixes multiple unrelated things.

Please

1. split this patch to multiple patches, each focused on single thing (e.g. PaX, tcl/tk, …),
2. add proper patch header with origin of the patch and brief explanation what does it do and why (if it’s not obvious from the patch itself),
3. fix mixed tabs and spaces.

File: testing/pypy3/const_stdio.patch [hunk 2]
```diff
@@ -0,0 +1,18 @@
+diff -r b30914a448be rpython/rlib/rfile.py
```
Comment (jirutka) [inline testing/pypy3/const_stdio.patch]: Why is this patch needed (please add patch header with description)? Have you sent it also to upstream, or is it specific for Alpine?

File: testing/pypy3/libpthread_version.patch [hunk 3]
```diff
@@ -0,0 +1,13 @@
+diff -r b30914a448be pypy/module/sys/system.py
```
Comment (jirutka) [inline testing/pypy3/libpthread_version.patch]: Why is this patch needed (please add patch header with description)? Have you sent it also to upstream, or is it specific for Alpine?

File: testing/pypy3/APKBUILD [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# Maintainer: Mihnea Saracin <mihnea.saracin@rinftech.com>
+pkgname=pypy3
+pkgver=5.7.1
+pkgrel=0
+pkgdesc="A fast, compliant alternative implementation of the Python 3 language"
+url="https://github.com/HorridBro/AlpinePyPy3"
+arch="all"
+license="MIT"
+depends=""
+makedepends="python linux-headers libffi-dev pkgconf zlib-dev bzip2-dev sqlite-dev ncurses-dev readline-dev libressl-dev expat-dev gdbm-dev paxmark tk tk-dev py3-cffi py2-cffi xz xz-dev"
```
Comment (jirutka) [inline testing/pypy3/APKBUILD]: There’s no package named `python`, please replace it with `python2`.

Wrap this very long line at ~80 chars.

File: testing/pypy3/APKBUILD [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+subpackages="$pkgname-dev"
+source="https://bitbucket.org/pypy/pypy/downloads/$pkgname-v$pkgver-src.tar.bz2
+        alpine.patch libpthread_version.patch const_stdio.patch"
```
Comment (jirutka) [inline testing/pypy3/APKBUILD]: Please stick to one item per line in `source`.

File: testing/pypy3/APKBUILD [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+builddir="$srcdir/$pkgname-v$pkgver-src"
+
+
+
```
Comment (jirutka) [inline testing/pypy3/APKBUILD]: One empty line between is enough, two are okay here too, but not three.

File: testing/pypy3/APKBUILD [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+build() {
+        cd "$builddir/pypy/goal"
+        python2 ../../rpython/bin/rpython --opt=jit --nopax targetpypystandalone.py 
+}
+
+
+
```
Comment (jirutka) [inline testing/pypy3/APKBUILD]: One empty line is enough.

File: testing/pypy3/APKBUILD [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+package() {
+        cd "$builddir/pypy/tool/release"
+        ./package.py --archive-name $pkgname-$pkgver --builddir "$pkgdir" 
+}
+
+check(){
+	cd "$builddir/lib-python/3/test/"
+	# XXX: Some tests fail because they are implemented using CPython C functions
+	$builddir/pypy/goal/pypy3-c regrtest.py -q || true
+}
```
Comment (jirutka) [inline testing/pypy3/APKBUILD]: `check` runs between `build` and `package`, so please put it between them, to make it more clear.

**Comment by jirutka** [comment]
There’s missed indentation in the APKBUILD, some lines are indented with spaces instead of tabs.

**Comment by ganwell** [comment]
Hi, sadly I missed this since I usually use the aports mailinglist. Here is my version of the package:

https://github.com/ganwell/aports/blob/pypy/testing/pypy3/APKBUILD

Please consider my smoketest

```bash
check() {
	local pypy3="$builddir/pypy/goal/pypy3-c"
	$pypy3 smoketest.py | grep "hello world"
	$pypy3 pystone.py
}
```

It helped me to detect the pax issues.

Also consider that the default split will move the .c files to -dev, but since pypy will compile some .c files at runtime, these actually belong to the main package. (I haven't tested your package, yet)

**Comment by HorridBro** [comment]
Thank you @ganwell. I will take a look at your repo and update my APKBUILD .

**Comment by ganwell** [comment]
@HorridBro First of all I'm a bit confused why some of the these things fail on my build chroot and not in yours, but I report them anyways.

## hardened

I tested the build, it fails on hardened

```
>>> pypy3: Entering fakeroot...
Got an unexpected error trying to allocate some memory for the JIT (tried to do mmap() with PROT_EXEC|PROT_READ|PROT_WRITE).  This can be caused by a system policy like PAX.  You need to find how to work around the policy on your system.
Aborted
>>> pypy3: Entering fakeroot...
Got an unexpected error trying to allocate some memory for the JIT (tried to do mmap() with PROT_EXEC|PROT_READ|PROT_WRITE).  This can be caused by a system policy like PAX.  You need to find how to work around the policy on your system.
Traceback (most recent call last):
  File "./package.py", line 335, in <module>
    retval, _ = package(*sys.argv[1:])
  File "./package.py", line 314, in package
    return create_package(basedir, options, **kwds)
  File "./package.py", line 84, in create_package
    raise OSError("Running %r failed!" % (str(pypy_c),))
OSError: Running '/home/ganwell/Repositories/aports/testing/pypy3/src/pypy3-v5.8.0-src/pypy/goal/pypy3-c' failed!
>>> ERROR: pypy3*: package failed
>>> ERROR: pypy3: all failed
```

In my version I set the attr after translation:

```
python2 ../../rpython/bin/rpython -Ojit --shared targetpypystandalone
attr -q -s pax.flags -V m ./pypy3-c
```

## libressl

```
In file included from /usr/include/openssl/x509.h:595:0,
                 from /usr/include/openssl/engine.h:96,
                 from _pypy_openssl.c:518:
_pypy_openssl.c:2747:19: error: expected identifier or '(' before numeric constant
 static const long X509_V_ERR_HOSTNAME_MISMATCH = 0;
                   ^
_pypy_openssl.c:2748:19: error: expected identifier or '(' before numeric constant
 static const long X509_V_ERR_EMAIL_MISMATCH = 0;
                   ^
_pypy_openssl.c:2749:19: error: expected identifier or '(' before numeric constant
 static const long X509_V_ERR_IP_ADDRESS_MISMATCH = 0;
```

Please see my patch https://github.com/ganwell/aports/blob/pypy/testing/pypy3/dont_redefine_mismatch_constants.patch

And my issue https://bitbucket.org/pypy/pypy/issues/2605/libress-defines-x509-__mismatch-constants

## tracing

```
>>> ERROR: pypy3*: libpypy3-c.so: path not found
        so:libbz2.so.1
        so:libc.musl-x86_64.so.1
        so:libcrypto.so.41
        so:libexpat.so.1
        so:libffi.so.6
        so:libgcc_s.so.1
        so:libgdbm.so.4
        so:liblzma.so.5
        so:libncursesw.so.6
        so:libpanelw.so.6
        so:libsqlite3.so.0
        so:libssl.so.43
        so:libtcl8.6.so
        so:libtk8.6.so
        so:libz.so.1
```

to fix this you need to move libpypy3-c.so

`mv "$pkgdir/usr/pypy3/bin/libpypy3-c.so" "$pkgdir/usr/lib/libpypy3-c.so"`

## Installation location

You install pypy into /pypy3-5.8.0? I would propose to install it into /usr/pypy3 as I do, but with your method of directly putting it there.

## lib_pypy

As soon as you install pypy to /usr this becomes relevant.

The the .c files in lib_pypy don't seem to be used in everyday pypy3, but when I ran some test-suites it didn't find the files, I still think lib_pypy shouldn't be split.

I am currently building and testing your pypy3 package.

## Final proposal

I would recommend to merge your and my APKBUILD since they both do some things better.

**Comment by HorridBro** [comment]
Ok, that sounds like a good idea. I`m open to any collaboration. How can we proceed with the merge?

**Comment by ganwell** [comment]
How about it try to merge my APKBUILD into yours? And you check the result and we fix remaining issues together?

I also like to ask some alpine maintainer about the resulting directory structure, since pypy is a bit unusual. I will post the structure after my next build (they take soooo long :smile:) Since you did PR 551 on pypy you definitely know more about the pypy side. I wasn't even sure on what branches I should do a PR, so I sent a bunch of patches. (And missed your PR :frowning:)

**Comment by jirutka** [comment]
Please don't set PaX flags directly, use paxmark script from the paxmark package.

**Comment by ganwell** [comment]
@jirutka Here the intended directory structure

* pypy3 actually proposes to be installed in /opt/pypy3 which is not allowed in alpine
* Therefore I install it in /usr/pypy3 and
  * Link /usr/pypy3/bin/pypy3 to /usr/bin/pypy3 as proposed by [build.rst](https://github.com/mozillazg/pypy/blob/master/pypy/doc/build.rst#installation) from pypy3
  * Move /usr/pypy3/bin/libpypy3-c.so to /pypy3/usr/lib/libpypy3-c.so: note there is no versioned so, since pypy is self-contained (also needed to fix so-tracing)

Other distros do exactly this except that they actually use /opt/pypy3 instead of /usr/pypy3

* Additionally I also link /usr/pypy3/include to /usr/include/pypy3

Are there any problem I missed?

Here the complete [structure](https://gist.github.com/ganwell/384c464be731e3e5c2f7f77a1048ad65)

Below a the shortened structure:

```
pkg/pypy3/usr/bin
pkg/pypy3/usr/bin/pypy3 (link to /usr/pypy3/bin/pypy3)
pkg/pypy3/usr/pypy3
pkg/pypy3/usr/pypy3/site-packages
pkg/pypy3/usr/pypy3/site-packages/README
pkg/pypy3/usr/pypy3/lib_pypy
.....
pkg/pypy3/usr/pypy3/lib_pypy/_syslog_cffi.pypy3-58-x86_64-linux-gnu.so
pkg/pypy3/usr/pypy3/lib_pypy/_lzma_cffi.pypy3-58-x86_64-linux-gnu.so
pkg/pypy3/usr/pypy3/lib_pypy/_codecs_jp.py
pkg/pypy3/usr/pypy3/lib_pypy/_marshal.py
pkg/pypy3/usr/pypy3/lib_pypy/_decimal.py
pkg/pypy3/usr/pypy3/lib_pypy/_gdbm_build.py
pkg/pypy3/usr/pypy3/lib_pypy/_audioop_build.py
pkg/pypy3/usr/pypy3/lib_pypy/_cffi_ssl
pkg/pypy3/usr/pypy3/lib_pypy/_cffi_ssl/LICENSE
pkg/pypy3/usr/pypy3/lib_pypy/_cffi_ssl/_cffi_src
pkg/pypy3/usr/pypy3/lib_pypy/_cffi_ssl/_cffi_src/utils.py
pkg/pypy3/usr/pypy3/lib_pypy/_cffi_ssl/_cffi_src/hazmat_src
.....
pkg/pypy3/usr/pypy3/lib_pypy/_cffi_ssl/tools/make_ssl_data.py
pkg/pypy3/usr/pypy3/lib_pypy/_structseq.py
pkg/pypy3/usr/pypy3/lib_pypy/_libmpdec
pkg/pypy3/usr/pypy3/lib_pypy/_libmpdec/fourstep.c
pkg/pypy3/usr/pypy3/lib_pypy/_libmpdec/crt.c
.....
pkg/pypy3/usr/pypy3/lib_pypy/_audioop_cffi.pypy3-58-x86_64-linux-gnu.so
.....
pkg/pypy3/usr/pypy3/lib_pypy/_ffi.py
pkg/pypy3/usr/pypy3/bin
pkg/pypy3/usr/pypy3/bin/pypy3
pkg/pypy3/usr/pypy3/lib-python
pkg/pypy3/usr/pypy3/lib-python/3
pkg/pypy3/usr/pypy3/lib-python/3/sunau.py
.....
pkg/pypy3/usr/pypy3/lib-python/3/plat-sunos5/TYPES.py
pkg/pypy3/usr/lib
pkg/pypy3/usr/lib/libpypy3-c.so (moved here from /usr/pypy3/bin/ for tracing)
pkg/pypy3-dev/usr/pypy3/include
pkg/pypy3-dev/usr/pypy3/include/sysmodule.h
pkg/pypy3-dev/usr/pypy3/include/pycobject.h
pkg/pypy3-dev/usr/pypy3/include/cpyext_unicodeobject.h
....
pkg/pypy3-dev/usr/include
pkg/pypy3-dev/usr/include/pypy3 (link to /usr/pypy3/include)
pkg/pypy3-doc/usr/share/doc/pypy3/LICENSE
pkg/pypy3-doc/usr/share/doc/pypy3/README.rst
```

**Comment by ganwell** [comment]
@jirutka I just compared to erlang and erlang is installed in /usr/lib/erlang, maybe we should install to /usr/lib/pypy3?

If so the above listing is still valid except /usr/pypy3 -> /usr/lib/pypy3.

**Comment by HorridBro** [comment]
@ganwell I`ve updated the APKBUILD, can you take a look?

**Comment by ganwell** [comment]
@HorridBro Great, thanks a lot! Looks very good. I started the build, to test it.

**Comment by ganwell** [comment]
@HorridBro It worked well: I tested some of my utility python-scripts, they all run with pypy3. I think that't it. :+1: I noticed that there are spaces in APKBUILD, Alpine uses tabs.

```
[APKBUILD]
indent_style = space
indent_size = 8
```

If your editor supports [editorconfig](http://editorconfig.org/).

@jirutka 
1. Are there remaining issues?
2. What about /usr/pypy3 vs /usr/lib/pypy3 (see above)

**Comment by HorridBro** [comment]
@ganwell , I fixed the tabs problem.

**Pull request closed without merging**